src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs

```diff
@@ -1,16 +1,935 @@
+    IL_00f6:  call       ""void System.Threading.Tasks.ManualResetValueTaskSourceLogic<bool>.Reset()""
+    IL_00fb:  nop
+    IL_00fc:  ldarg.0
+    IL_00fd:  ldflda     ""System.Threading.Tasks.ManualResetValueTaskSourceLogic<bool> C.<M>d__0.<>v__promiseOfValueOrEnd""
+    IL_0102:  ldc.i4.0
+    IL_0103:  call       ""void System.Threading.Tasks.ManualResetValueTaskSourceLogic<bool>.SetResult(bool)""
+    IL_0108:  nop
+    IL_0109:  leave.s    IL_012f
+  }
+  catch System.Exception
+  {
+    // sequence point: <hidden>
+    IL_010b:  stloc.3
+    IL_010c:  ldarg.0
+    IL_010d:  ldc.i4.s   -2
+    IL_010f:  stfld      ""int C.<M>d__0.<>1__state""
+    IL_0114:  ldarg.0
+    IL_0115:  ldfld      ""bool C.<M>d__0.<>w__promiseIsActive""
+    IL_011a:  brfalse.s  IL_012b
+    IL_011c:  ldarg.0
+    IL_011d:  ldflda     ""System.Threading.Tasks.ManualResetValueTaskSourceLogic<bool> C.<M>d__0.<>v__promiseOfValueOrEnd""
+    IL_0122:  ldloc.3
+    IL_0123:  call       ""void System.Threading.Tasks.ManualResetValueTaskSourceLogic<bool>.SetException(System.Exception)""
+    IL_0128:  nop
+    IL_0129:  br.s       IL_012d
+    IL_012b:  rethrow
+    IL_012d:  leave.s    IL_012f
+  }
+  IL_012f:  ret
+}", sequencePoints: "C+<M>d__0.MoveNext", source: source);
+        }
+
+        [Fact]
+        public void AsyncIteratorWithReturn()
+        {
+            string source = @"
+class C
+{
+    public static async System.Collections.Generic.IAsyncEnumerable<int> M()
+    {
+        await System.Threading.Tasks.Task.CompletedTask;
+        yield return 3;
+        return null;
+    }
+}";
+            var comp = CreateCompilationWithTasksExtensions(source + s_common);
+            comp.VerifyDiagnostics(
+                // (8,9): error CS1622: Cannot return a value from an iterator. Use the yield return statement to return a value, or yield break to end the iteration.
+                //         return null;
+                Diagnostic(ErrorCode.ERR_ReturnInIterator, "return").WithLocation(8, 9)
+                );
+        }
+
+        [Fact]
+        public void AsyncIteratorWithAwaitCompletedAndTwoYields()
+        {
+            string source = @"
+using static System.Console;
+class C
+{
+    static async System.Collections.Generic.IAsyncEnumerable<int> M()
+    {
+        Write(""1 "");
+        await System.Threading.Tasks.Task.CompletedTask;
+        Write(""2 "");
+        yield return 3;
+        Write(""4 "");
+        yield return 5;
+    }
+    static async System.Threading.Tasks.Task Main()
+    {
+        Write(""0 "");
+        foreach await (var i in M())
+        {
+            Write($""{i} "");
+        }
+        Write(""Done"");
+    }
+}";
+            var comp = CreateCompilationWithTasksExtensions(source + s_common, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "0 1 2 3 4 5 Done", verify: Verification.Passes);
```

Comment:
>CompileAndVerify(comp, expectedOutput: "0 1 2 3 4 5 Done", verify: Verification.Passes); [](start = 12, length = 88)

nit: all of these tests could be simplified into `CompileAndVerify(source + s_common, expectedOutput: "0 1 2 3 4 5 Done");`
This will verify both IL and errors.

Reply:
`VerifyDiagnostics` also verifies warnings, which seems useful.